merkle/tests/tree.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+// Test cases are generated from ocaml code
+//
+// #use "https://github.com/o1-labs/snarky/blob/master/src/base/merkle_tree.ml"
+//
+// let my_hash x = x
+
+// let my_merge a b =
+//   match a with
+//   | None -> None
+//   | Some a -> ( match b with None -> Some a | Some b -> Some (a + b))
+
+// let create_tree n =
+//   let tree = create ~hash:my_hash ~merge:my_merge 0 in
+//   add_many tree (List.init (n - 1) ~f:(fun i -> i + 1))
+//
+// create_tree 127 |> root
+// create_tree 127 |> depth
+
+#[cfg(test)]
+mod tests {
+    use mina_merkle::*;
+
+    struct TestHasher {}
+
+    impl MerkleHasher for TestHasher {
+        type Item = i64;
+        type Hash = i64;
+        fn hash(item: &Self::Item, _: MerkleTreeNodeMetadata) -> Self::Hash {
+            *item
+        }
+    }
+
+    struct TestMerger {}
+
+    impl MerkleMerger for TestMerger {
+        type Hash = i64;
+        fn merge(
+            left: &Option<(Self::Hash, MerkleTreeNodeMetadata)>,
+            right: &Option<(Self::Hash, MerkleTreeNodeMetadata)>,
+        ) -> Option<Self::Hash> {
+            if let Some((left, _)) = left {
```

Comment:
this could be written like this:
```rust
left.as_ref().and_then(|(left, _)| right.as_ref().and_then(|(right, _)| Some(left + right)))
```

but not sure if that's more readable.

Reply:
Tried this but tests failed, looks like this does not handle line 48 tho,

Reply:
oh that can be handled with chaining `or` after right's and_then:
```rust
left.as_ref().and_then(|(left, _)| right.as_ref().and_then(|(right, _)| Some(left + right)).or(Some(*left)))
```

but i guess it's okay to keep it the usual if else way.